query-grammar/src/query_grammar.rs

```diff
@@ -413,38 +735,121 @@ fn aggregate_binary_expressions(
     }
 }
 
-fn operand_leaf<'a>() -> impl Parser<&'a str, Output = (BinaryOperand, UserInputAst)> {
-    (
-        binary_operand().skip(spaces()),
-        boosted_leaf().skip(spaces()),
-    )
+fn aggregate_infallible_expressions(
+    leafs: Vec<(Option<BinaryOperand>, Option<Occur>, Option<UserInputAst>)>,
+) -> (UserInputAst, ErrorList) {
+    let mut leafs: Vec<(_, _, UserInputAst)> = leafs
+        .into_iter()
+        .filter_map(|(operand, occur, ast)| ast.map(|ast| (operand, occur, ast)))
+        .collect();
+    if leafs.is_empty() {
+        // TODO should this produce an error?
+        return (UserInputAst::empty_query(), Vec::new());
+    }
+
+    // if no occur => use binary operand (default = or)
+    // if no binary opearnd => use occur
+
+    if leafs.len() == 1 {
+        match leafs.remove(0) {
+            (_, occur, ast) if occur != Some(Occur::MustNot) => return (ast, Vec::new()),
+            (_, occur, ast) => return (UserInputAst::Clause(vec![(occur, ast)]), Vec::new()),
+        }
+    }
+
+    if leafs.iter().all(|(operand, _, _)| operand.is_none()) {
+        let clauses = leafs
+            .into_iter()
+            .map(|(_, occur, ast)| (occur, ast))
+            .collect();
+        return (UserInputAst::Clause(clauses), Vec::new());
+    }
+
+    if leafs.iter().all(|(_, occur, _)| occur.is_none()) {
+        let first = leafs.remove(0).2;
+        let rest = leafs
+            .into_iter()
+            .map(|(operand, _, ast)| (operand.unwrap_or(BinaryOperand::Or), ast))
```

Comment:
should the operand depend on the query parser default behavior?

Reply:
this part needs to be improved to work with a mix of binary operand and occurs, however I'm yet to find how to mix both in a way that makes sens and transition coherently from full-occurs and full-binary-operand